### PR TITLE
Charity vaults factory

### PR DESCRIPTION
**Pull request by abigger87**
## Abstract

This pr introduces the `CharityVaultFactory` contract. It also completely rewrites the `CharityVault` contract to wrap the `Vault` contract - abstracting away the underlying Vault logic, isolating charity logic to this contract.

Respective test files can be found in `src/tests/*.t.sol`

## Progress

- [x] CharityVaultFactory
    - [x] deployCharityVault
    - [x] getCharityVaultFromUnderlying
    - [x] isCharityVaultDeployed
- [ ] CharityVaultFactory.t.sol
    - [x] test_basic_charity_vault_not_deployed 
    - [x] test_able_to_deploy_charity_vault_from_factory
- [ ] CharityVault.sol
- [ ] CharityVault.t.sol
- [ ]

### CODE REVIEW

File: src/CharityVault.sol [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    // ?? Either the CharityVault is a Vault ??
+    // ?? Or, it deploys/finds an existing Vault and wraps it managing the fvTokens <> fcvTokens ??
+
+
+    /*///////////////////////////////////////////////////////////////
+                                CONSTANTS
+    //////////////////////////////////////////////////////////////*/
+
+    address payable charity;
+    uint256 feePercent;
```
Comment (transmissions11) [inline src/CharityVault.sol:30]: make immutable?

File: src/CharityVaultFactory.sol [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    /// @return cvault The newly deployed CharityVault contract.
+    function deployCharityVault(ERC20 underlying, address payable charity, uint256 feePercent) external returns (CharityVault cvault) {
+        // Compute a CharityVault Salt
+        bytes32 cvaultSalt = keccak256(
+            abi.encodePacked(
+                address(underlying),
+                charity,
+                feePercent
+            )
+        );
```
Comment (transmissions11) [inline src/CharityVaultFactory.sol:41]: why not just inline this for the salt arg?

File: src/CharityVaultFactory.sol [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    /// @return The CharityVault that supports this underlying token.
+    function getCharityVaultFromUnderlying(ERC20 underlying, address payable charity, uint256 feePercent) external view returns (CharityVault) {
+        // Compute a CharityVault's Salt
+        bytes32 cvaultSalt = keccak256(
+            abi.encodePacked(
+                address(underlying),
+                charity,
+                feePercent
+            )
+        );
```
Comment (transmissions11) [inline src/CharityVaultFactory.sol:69]: why not inline?

File: src/CharityVault.sol (hunk 1 above)
Comment (abigger87) [reply src/CharityVault.sol:30]: oo cool didn't know u could do that

File: src/CharityVaultFactory.sol (hunk 2 above)
Comment (abigger87) [reply src/CharityVaultFactory.sol:41]: inline?

Comment (transmissions11) [reply src/CharityVaultFactory.sol:41]: like remove this variable and replace uses of `cvaultSalt` with 

```
keccak256(
            abi.encodePacked(
                address(underlying),
                charity,
                feePercent
            )
        )
``` 

that way u dont waste gas loading it into memory and then reading from it only once

File: src/tests/CharityVaultFactory.t.sol [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+    function setUp() public {
+        underlying = new MockERC20("Mock Token", "TKN", 18);
+        factory = new CharityVaultFactory();
+    }
+
+
+    // The CharityVaultFactory should not be able to find a deployed vault for anything initially
+    function test_basic_charity_vault_not_deployed(address fuzzed_addr) public {
+        assertTrue(
```
Comment (transmissions11) [inline src/tests/CharityVaultFactory.t.sol:22]: DSTestPlus provides `assertFalse`

**Review by transmissions11 — APPROVED** [review]

**Comment by transmissions11** [comment]
overall lgtm— one thing is that i'd remove package-lock.json if u want this to be a yarn project, and update the makefile to install with yarn as well

File: src/CharityVaultFactory.sol [hunk 5]
```diff
@@ -0,0 +1,103 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.6;
+
+import {ERC20} from "solmate/erc20/ERC20.sol";
+import {Bytes32AddressLib} from "./libraries/Bytes32AddressLib.sol";
+import "./CharityVault.sol";
```
Comment (transmissions11) [inline src/CharityVaultFactory.sol:6]: oh can we make this a named import like above pls dont like leaking other contracts into scope

File: src/CharityVault.sol [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+pragma solidity 0.8.6;
+
+import {ERC20} from "solmate/erc20/ERC20.sol";
+import {Auth} from "solmate/auth/Auth.sol";
+import {SafeERC20} from "solmate/erc20/SafeERC20.sol";
+
+import {WETH} from "./external/WETH.sol";
+import {CErc20} from "./external/CErc20.sol";
+
+import "./tests/utils/DSTestPlus.sol";
```
Comment (transmissions11) [inline src/CharityVault.sol:11]: dont think u need this anymore hehe

File: src/tests/CharityVaultFactory.t.sol [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+            !factory.isVaultDeployed(CharityVault(payable(fuzzed_addr)))
+        );
+    }
+
+    /// @notice we can deploy a CharityVault
+    /// @dev uses the same testing pattern as VaultFactory for consistency
+    function test_able_to_deploy_charity_vault_from_factory(address fuzzed_addr, uint256 feePercent) public {
+      // Validate fuzzed fee percent
+      if (feePercent > type(uint256).max / 1e36) return;
```
Comment (transmissions11) [inline src/tests/CharityVaultFactory.t.sol:31]: hm doesn't look like this bound is enforced in the contracts anywhere? if u don't want to accept fees over a certain % mayb validate in the charity vault constructor

File: src/CharityVault.sol (hunk 6 above)
Comment (abigger87) [reply src/CharityVault.sol:11]: Oh I thought we decided the `CharityVault` will mint and burn another token the `fcvToken`. But yea current code is not updated. Working on CharityVault right now.

Comment (transmissions11) [reply src/CharityVault.sol:11]: wot? just talking about importing DSTestPlus

**Comment by abigger87** [comment]
Sending a merge so others can work on this - cadence should be frequent merges into develop for syncronicity

**Pull request merged**